### PR TITLE
Implement bookmarks

### LINKED ISSUE #102: Feature to bookmark articles
The app could add the ability to bookmark articles. This could be useful in multiple ways. You could use it as a "read-later" of sorts or you could use it to save really good articles. 

There could be an additional tab which has all bookmarked articles.

### PR DESCRIPTION
**Pull request by ZacSweers**
Resolves #102

Some highlights
- Using a nested Circuit screen in a list item for the bookmarks UI
- Separate composite bookmarks DB
- A little more compose KMP work
- New Bookmarks screen with some swipe to dismiss

https://github.com/ZacSweers/CatchUp/assets/1361086/7eed0c36-ac76-4be2-9019-80c879aa0f60

### CODE REVIEW

**Comment by ZacSweers** [comment]
I should add an export feature

File: bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  // Exposed to create a PagingSource
+  fun bookmarksCountQuery(): Query<Long>
+
+  fun bookmarksTransacter(): Transacter
+
+  fun bookmarksQuery(limit: Long, offset: Long): Query<CatchUpItem>
+}
+
+internal class BookmarkRepositoryImpl(private val database: CatchUpDatabase) : BookmarkRepository {
+  private val scope = CoroutineScope(Dispatchers.IO)
```
Comment (jingibus) [inline bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt:40]: Inject this. Constructing this scope internally bakes in the assumption that the class is singleton-y; not doing this is a major reason we use the DI pattern and DI frameworks like Dagger.

Also, don't inject as a `val`. See comments below about launches on instance methods. Launches on initialization are fine IMO, and don't require making the scope a `val`, either.

File: bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        database.transactionWithResult { database.bookmarksQueries.bookmarkIds().asFlow() }
+      idsFlow.collect { query ->
+        // Preserve order
+        bookmarks.emit(query.executeAsList().mapTo(LinkedHashSet(), Bookmark::id))
+      }
+    }
+  }
+
+  override fun addBookmark(id: Long, timestamp: Instant) {
+    scope.launch { database.transaction { database.bookmarksQueries.addBookmark(id, timestamp) } }
```
Comment (jingibus) [inline bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt:57]: This API probably wants to be a `suspend fun`! In many scenarios the caller can and should safely own this, and a `suspend fun` still opens the door for kicking it over to a longer lived scope if the caller really does need that

If you feel strongly opinionated that this is the right API shape, then `scope.launch` *still* isn't the right choice. Use a channel instead:

```kotlin
data class AddBookmark(val id: Long, val timestamp: Instant)

private val addBookmarks = Channel<AddBookmark>(BUFFERED)

init {
  scope.launch {
    for (item in addBookmarks) {
      database.transaction { database.bookmarksQueries.addBookmark(item.id, item.timestamp) }
    }
  }
}

override fun addBookmark(id: Long, timestamp: Instant) {
  addBookmarks.sendOrFail(AddBookMark(id, timestamp))
}
```

You might say, "I don't need a queue! Why did you make me build a queue?" But you already had a queue: it was the dispatcher, and its failure mode was "my dispatcher is full of hung bookmark add operations".

File: bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  override fun removeBookmark(id: Long) {
+    scope.launch { database.transaction { database.bookmarksQueries.removeBookmark(id) } }
```
Comment (jingibus) [inline bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt:61]: Same

File: bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  init {
+    scope.launch {
+      val idsFlow =
+        database.transactionWithResult { database.bookmarksQueries.bookmarkIds().asFlow() }
+      idsFlow.collect { query ->
+        // Preserve order
+        bookmarks.emit(query.executeAsList().mapTo(LinkedHashSet(), Bookmark::id))
+      }
+    }
+  }
```
Comment (jingibus) [inline bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt:54]:
```suggestion
private val bookmarks = database.transactionWithResult { database.bookmarksQueries.bookmarkIds().asFlow() }
  .map { it.executeAsList().mapTo(LinkedHashSet(), Bookmark::id) }
  .stateIn(scope, SharingStarted.Eagerly, LinkedHashSet<Long>())
```

File: bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt (hunk 2 above)
Comment (ZacSweers) [reply bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt:57]: my concern about making it suspend is that I don't fully understand what that communicates to the caller. What happens if they cancel the add? Is that something the UI gets to really decide? Does the add() always work but the UI gets to know if it successfully handed it off (synchronously) to the internal transaction handler?

Comment (jingibus) [reply bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt:57]: * When the `suspend fun` write completes successfully, the caller knows the write succeeded
* If the write fails, the caller will get an exception (which is good)
* If the caller cancels the `addBookmark` call before it completes (e.g. by navigating away), then the write will get discarded
* If the caller wants to navigate away and not cancel the write, they can still accomplish that by doing exactly what your existing code here does

All of these are valid choices, which is why I tend to prefer suspend funs for this type of thing.

**Comment by ZacSweers** [comment]
Thanks @jingibus!

**Pull request merged**